File: apps/devhud/native-widgets/android/shared/src/main/java/dev/deli/devhud/widget/WidgetConfiguration.kt
package dev.deli.devhud.widget

import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.UUID
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

object DevHudWidgetContract {
    const val EXTENSION_IDENTIFIER = "dev.deli.devhud.widget"
    const val STORAGE_KEY = "devhud.widget-configuration.v1"
    const val DATASTORE_NAME = "devhud-widget-state"
    const val DATASTORE_FILE = "$DATASTORE_NAME.preferences_pb"
    const val KEY_ALIAS_PREFIX = "devhud.deck-widget.snapshot.v1."
    const val SCHEMA_VERSION = 1
    const val MAX_WIDGETS = 20
    const val MAX_PULL_REQUESTS = 10
    const val GENERIC_NOTIFICATION_TEXT = "Deck view updated"
    const val APP_LINK_PATH = "/devhud/deck/open"
}

enum class DeckWidgetFamily(val wireValue: String) {
    APPLE_SMALL("apple-small"), APPLE_MEDIUM("apple-medium"), APPLE_LARGE("apple-large"),
    ANDROID_COMPACT("android-compact"), ANDROID_WIDE("android-wide"), ANDROID_LIST("android-list");
    companion object {
        fun fromWireValue(value: String) = entries.firstOrNull { it.wireValue == value }
    }
}

enum class DeckWidgetPrivacy(val wireValue: String) {
    COUNTS_ONLY("counts-only"), REPOSITORY_AND_TITLES("repository-and-titles");
    companion object {
        fun fromWireValue(value: String) = entries.firstOrNull { it.wireValue == value }
    }
}

enum class DeckWidgetFreshness(val wireValue: String) {
    FRESH("fresh"), STALE("stale"), OFFLINE("offline"), DISCONNECTED("disconnected"),
    NEVER_REFRESHED("never-refreshed");
    companion object {
        fun fromWireValue(value: String) = entries.firstOrNull { it.wireValue == value }
    }
}

data class WidgetPullRequest(
    val repositoryOwner: String,
    val repositoryName: String,
    val number: Long,
    val title: String,
)

data class WidgetSnapshot(
    val matchingCount: Int,
    val pullRequests: List<WidgetPullRequest>,
    val freshness: DeckWidgetFreshness,
    val offline: Boolean,
    val generatedAt: String,
)

data class DeckWidgetInstance(
    val widgetId: String,
    val viewId: String,
    val family: DeckWidgetFamily,
    val privacy: DeckWidgetPrivacy,
    val snapshot: WidgetSnapshot,
)

data class WidgetConfiguration(val accountId: String, val widgets: List<DeckWidgetInstance>)

data class WidgetConfigurationRecord(val version: Int, val configuration: WidgetConfiguration) {
    companion object {
        val EMPTY = WidgetConfigurationRecord(
            DevHudWidgetContract.SCHEMA_VERSION,
            WidgetConfiguration("", emptyList()),
        )
    }
}

enum class WidgetConfigurationErrorCode(val wireValue: String) {
    CORRUPT("corrupt"), FUTURE_VERSION("future-version"), INCOMPATIBLE("incompatible"),
    REFRESH_FAILED("refresh-failed"), STORAGE_UNAVAILABLE("storage-unavailable"),
    WRITE_FAILED("write-failed"), ENCRYPTION_FAILED("encryption-failed"),
}

class WidgetConfigurationException(
    val code: WidgetConfigurationErrorCode,
    cause: Throwable? = null,
) : Exception(code.wireValue, cause)

object WidgetConfigurationCodec {
    fun decode(raw: String): WidgetConfigurationRecord {
        val root = try { JSONObject(raw) } catch (error: JSONException) {
            throw WidgetConfigurationException(WidgetConfigurationErrorCode.CORRUPT, error)
        }
        exactKeys(root, setOf("version", "configuration"))
        val version = integer(root.opt("version")) ?: incompatible()
        if (version > DevHudWidgetContract.SCHEMA_VERSION) {
            throw WidgetConfigurationException(WidgetConfigurationErrorCode.FUTURE_VERSION)
        }
        if (version != DevHudWidgetContract.SCHEMA_VERSION) incompatible()
        val configuration = root.opt("configuration") as? JSONObject ?: incompatible()
        exactKeys(configuration, setOf("accountId", "widgets"))
        val accountId = configuration.opt("accountId") as? String ?: incompatible()
        val array = configuration.opt("widgets") as? JSONArray ?: incompatible()
        if (array.length() > DevHudWidgetContract.MAX_WIDGETS) incompatible()
        if (array.length() > 0 && !validUuid(accountId)) incompatible()
        val widgets = mutableListOf<DeckWidgetInstance>()
        val widgetIds = mutableSetOf<String>()
        for (index in 0 until array.length()) {
            val value = array.opt(index) as? JSONObject ?: incompatible()
            exactKeys(value, setOf("widgetId", "viewId", "family", "privacy", "snapshot"))
            val widgetId = value.opt("widgetId") as? String ?: incompatible()
            val viewId = value.opt("viewId") as? String ?: incompatible()
            if (!validUuid(widgetId) || !validUuid(viewId) || !widgetIds.add(widgetId)) incompatible()
            val family = (value.opt("family") as? String)?.let(DeckWidgetFamily::fromWireValue) ?: incompatible()
            val privacy = (value.opt("privacy") as? String)?.let(DeckWidgetPrivacy::fromWireValue) ?: incompatible()
            val snapshot = decodeSnapshot(value.opt("snapshot") as? JSONObject ?: incompatible(), privacy)
            widgets += DeckWidgetInstance(widgetId, viewId, family, privacy, snapshot)
        }
        return WidgetConfigurationRecord(version, WidgetConfiguration(accountId, widgets))
    }

    fun encode(record: WidgetConfigurationRecord): String {
        val widgets = JSONArray()
        record.configuration.widgets.forEach { widget ->
            val pullRequests = JSONArray()
            widget.snapshot.pullRequests.forEach { pullRequest ->
                pullRequests.put(JSONObject()
                    .put("repositoryOwner", pullRequest.repositoryOwner)
                    .put("repositoryName", pullRequest.repositoryName)
                    .put("number", pullRequest.number)
                    .put("title", pullRequest.title))
            }
            widgets.put(JSONObject()
                .put("widgetId", widget.widgetId)
                .put("viewId", widget.viewId)
                .put("family", widget.family.wireValue)
                .put("privacy", widget.privacy.wireValue)
                .put("snapshot", JSONObject()
                    .put("matchingCount", widget.snapshot.matchingCount)
                    .put("pullRequests", pullRequests)
                    .put("freshness", widget.snapshot.freshness.wireValue)
                    .put("offline", widget.snapshot.offline)
                    .put("generatedAt", widget.snapshot.generatedAt)))
        }
        return JSONObject().put("version", record.version).put(
            "configuration",
            JSONObject().put("accountId", record.configuration.accountId).put("widgets", widgets),
        ).toString().also(::decode)
    }

    private fun decodeSnapshot(value: JSONObject, privacy: DeckWidgetPrivacy): WidgetSnapshot {
        exactKeys(value, setOf("matchingCount", "pullRequests", "freshness", "offline", "generatedAt"))
        val count = integer(value.opt("matchingCount")) ?: incompatible()
        if (count < 0) incompatible()
        val array = value.opt("pullRequests") as? JSONArray ?: incompatible()
        if (array.length() > DevHudWidgetContract.MAX_PULL_REQUESTS ||
            privacy == DeckWidgetPrivacy.COUNTS_ONLY && array.length() != 0) incompatible()
        val pullRequests = mutableListOf<WidgetPullRequest>()
        for (index in 0 until array.length()) {
            val pullRequest = array.opt(index) as? JSONObject ?: incompatible()
            exactKeys(pullRequest, setOf("repositoryOwner", "repositoryName", "number", "title"))
            val owner = pullRequest.opt("repositoryOwner") as? String ?: incompatible()
            val repository = pullRequest.opt("repositoryName") as? String ?: incompatible()
            val number = integer(pullRequest.opt("number")) ?: incompatible()
            val title = pullRequest.opt("title") as? String ?: incompatible()
            if (owner.isEmpty() || repository.isEmpty() || number <= 0 || title.isEmpty()) incompatible()
            pullRequests += WidgetPullRequest(owner, repository, number.toLong(), title)
        }
        val freshness = (value.opt("freshness") as? String)?.let(DeckWidgetFreshness::fromWireValue) ?: incompatible()
        val offline = value.opt("offline") as? Boolean ?: incompatible()
        val generatedAt = value.opt("generatedAt") as? String ?: incompatible()
        try { Instant.parse(generatedAt) } catch (_: Exception) { incompatible() }
        return WidgetSnapshot(count, pullRequests, freshness, offline, generatedAt)
    }

    private fun integer(value: Any?): Int? =
        (value as? Number)?.takeIf { it.toDouble() % 1.0 == 0.0 }?.toInt()
    private fun validUuid(value: String) = try { UUID.fromString(value); true } catch (_: Exception) { false }
    private fun exactKeys(value: JSONObject, expected: Set<String>) {
        if (value.keys().asSequence().toSet() != expected) incompatible()
    }
    private fun incompatible(): Nothing = throw WidgetConfigurationException(WidgetConfigurationErrorCode.INCOMPATIBLE)
}

sealed interface DeckWidgetAction {
    val viewId: String
    data class OpenView(override val viewId: String) : DeckWidgetAction
    data class OpenPullRequest(
        override val viewId: String,
        val owner: String,
        val repository: String,
        val number: Long,
    ) : DeckWidgetAction
    data class Refresh(override val viewId: String) : DeckWidgetAction
    data class ResolveEvent(val eventId: String) : DeckWidgetAction { override val viewId = "" }
}

fun DeckWidgetAction.toAppLink(): String {
    val values = mutableListOf(
        "action" to when (this) {
            is DeckWidgetAction.OpenView -> "open-view"
            is DeckWidgetAction.OpenPullRequest -> "open-pr"
            is DeckWidgetAction.Refresh -> "refresh"
            is DeckWidgetAction.ResolveEvent -> "resolve-event"
        },
    )
    if (this is DeckWidgetAction.ResolveEvent) values += "event" to eventId
    else values += "view" to viewId
    if (this is DeckWidgetAction.OpenPullRequest) {
        values += listOf("owner" to owner, "repository" to repository, "number" to number.toString())
    }
    val query = values.joinToString("&") { (key, value) ->
        "${encode(key)}=${encode(value)}"
    }
    return "https://deli.dev/devhud/deck/open?$query"
}

private fun encode(value: String) = URLEncoder.encode(value, StandardCharsets.UTF_8.toString())

object DeckNotificationPolicy {
    private val opaqueEvent = Regex("^[A-Za-z0-9_-]{16,128}$")
    fun eventId(payload: Map<String, String>): String? =
        payload["eventId"]?.takeIf { payload.keys == setOf("eventId") && opaqueEvent.matches(it) }
    fun text(detailedText: String?, localDetailEnabled: Boolean): String =
        detailedText?.takeIf { localDetailEnabled && it.isNotEmpty() }
            ?: DevHudWidgetContract.GENERIC_NOTIFICATION_TEXT
}

File: apps/devhud/src-tauri/widget-bridge/ios/Sources/DevHudWidgetCore/WidgetConfiguration.swift
import CoreFoundation
import Foundation

public enum DevHudWidgetContract {
    public static let appGroupIdentifier = "group.dev.deli.devhud"
    public static let extensionIdentifier = "dev.deli.devhud.widget"
    public static let storageKey = "devhud.widget-configuration.v1"
    public static let encryptionKeyService = "dev.deli.devhud.widget.snapshot-key.v1"
    public static let schemaVersion = 1
    public static let maximumWidgets = 20
    public static let maximumPullRequests = 10
    public static let genericNotificationText = "Deck view updated"
    public static let appLinkPath = "/devhud/deck/open"
}

public enum DeckWidgetFamily: String, CaseIterable, Codable, Sendable {
    case appleSmall = "apple-small"
    case appleMedium = "apple-medium"
    case appleLarge = "apple-large"
    case androidCompact = "android-compact"
    case androidWide = "android-wide"
    case androidList = "android-list"
}

public enum DeckWidgetPrivacy: String, CaseIterable, Codable, Sendable {
    case countsOnly = "counts-only"
    case repositoryAndTitles = "repository-and-titles"
}

public enum DeckWidgetFreshness: String, CaseIterable, Codable, Sendable {
    case fresh
    case stale
    case offline
    case disconnected
    case neverRefreshed = "never-refreshed"
}

public struct WidgetPullRequest: Codable, Equatable, Sendable {
    public let repositoryOwner: String
    public let repositoryName: String
    public let number: UInt64
    public let title: String
}

public struct WidgetSnapshot: Codable, Equatable, Sendable {
    public let matchingCount: UInt32
    public let pullRequests: [WidgetPullRequest]
    public let freshness: DeckWidgetFreshness
    public let offline: Bool
    public let generatedAt: String
}

public struct DeckWidgetInstance: Codable, Equatable, Sendable, Identifiable {
    public var id: String { widgetId }
    public let widgetId: String
    public let viewId: String
    public let family: DeckWidgetFamily
    public let privacy: DeckWidgetPrivacy
    public let snapshot: WidgetSnapshot
}

public struct WidgetConfiguration: Codable, Equatable, Sendable {
    public let accountId: String
    public let widgets: [DeckWidgetInstance]
}

public struct WidgetConfigurationRecord: Codable, Equatable, Sendable {
    public let version: Int
    public let configuration: WidgetConfiguration

    public static var empty: WidgetConfigurationRecord {
        WidgetConfigurationRecord(
            version: DevHudWidgetContract.schemaVersion,
            configuration: WidgetConfiguration(accountId: "", widgets: [])
        )
    }
}

public enum WidgetConfigurationError: String, Error, Equatable, Sendable {
    case corrupt
    case futureVersion = "future-version"
    case incompatible
    case refreshFailed = "refresh-failed"
    case sharedStoreUnavailable = "shared-store-unavailable"
    case writeFailed = "write-failed"
    case encryptionFailed = "encryption-failed"
}

public enum WidgetConfigurationCodec {
    public static func decode(_ data: Data) throws -> WidgetConfigurationRecord {
        let json: Any
        do { json = try JSONSerialization.jsonObject(with: data) }
        catch { throw WidgetConfigurationError.corrupt }
        guard let root = json as? [String: Any] else {
            throw WidgetConfigurationError.corrupt
        }
        try exactKeys(root, ["version", "configuration"])
        guard let version = integer(root["version"]) else {
            throw WidgetConfigurationError.incompatible
        }
        if version > DevHudWidgetContract.schemaVersion {
            throw WidgetConfigurationError.futureVersion
        }
        guard version == DevHudWidgetContract.schemaVersion,
              let configuration = root["configuration"] as? [String: Any]
        else { throw WidgetConfigurationError.incompatible }
        try validate(configuration)
        do { return try JSONDecoder().decode(WidgetConfigurationRecord.self, from: data) }
        catch { throw WidgetConfigurationError.incompatible }
    }

    public static func encode(_ record: WidgetConfigurationRecord) throws -> Data {
        let data = try JSONEncoder.sorted.encode(record)
        _ = try decode(data)
        return data
    }

    private static func validate(_ configuration: [String: Any]) throws {
        try exactKeys(configuration, ["accountId", "widgets"])
        guard let accountId = configuration["accountId"] as? String,
              let widgets = configuration["widgets"] as? [[String: Any]],
              widgets.count <= DevHudWidgetContract.maximumWidgets,
              widgets.isEmpty || UUID(uuidString: accountId) != nil
        else { throw WidgetConfigurationError.incompatible }
        var widgetIds = Set<String>()
        for widget in widgets {
            try exactKeys(widget, ["widgetId", "viewId", "family", "privacy", "snapshot"])
            guard let widgetId = widget["widgetId"] as? String,
                  UUID(uuidString: widgetId) != nil,
                  widgetIds.insert(widgetId).inserted,
                  let viewId = widget["viewId"] as? String,
                  UUID(uuidString: viewId) != nil,
                  let family = widget["family"] as? String,
                  DeckWidgetFamily(rawValue: family) != nil,
                  let privacyRaw = widget["privacy"] as? String,
                  let privacy = DeckWidgetPrivacy(rawValue: privacyRaw),
                  let snapshot = widget["snapshot"] as? [String: Any]
            else { throw WidgetConfigurationError.incompatible }
            try validate(snapshot, privacy: privacy)
        }
    }

    private static func validate(
        _ snapshot: [String: Any],
        privacy: DeckWidgetPrivacy
    ) throws {
        try exactKeys(snapshot, ["matchingCount", "pullRequests", "freshness", "offline", "generatedAt"])
        guard let count = integer(snapshot["matchingCount"]), count >= 0,
              let pullRequests = snapshot["pullRequests"] as? [[String: Any]],
              pullRequests.count <= DevHudWidgetContract.maximumPullRequests,
              privacy != .countsOnly || pullRequests.isEmpty,
              let freshness = snapshot["freshness"] as? String,
              DeckWidgetFreshness(rawValue: freshness) != nil,
              snapshot["offline"] is Bool,
              let generatedAt = snapshot["generatedAt"] as? String,
              ISO8601DateFormatter().date(from: generatedAt) != nil
        else { throw WidgetConfigurationError.incompatible }
        for pullRequest in pullRequests {
            try exactKeys(pullRequest, ["repositoryOwner", "repositoryName", "number", "title"])
            guard let owner = pullRequest["repositoryOwner"] as? String, !owner.isEmpty,
                  let repository = pullRequest["repositoryName"] as? String, !repository.isEmpty,
                  let number = integer(pullRequest["number"]), number > 0,
                  let title = pullRequest["title"] as? String, !title.isEmpty
            else { throw WidgetConfigurationError.incompatible }
        }
    }

    private static func integer(_ value: Any?) -> Int? {
        guard let number = value as? NSNumber,
              CFGetTypeID(number) != CFBooleanGetTypeID(),
              number.doubleValue.rounded() == number.doubleValue
        else { return nil }
        return number.intValue
    }

    private static func exactKeys(_ object: [String: Any], _ expected: Set<String>) throws {
        guard Set(object.keys) == expected else {
            throw WidgetConfigurationError.incompatible
        }
    }
}

private extension JSONEncoder {
    static var sorted: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }
}

public enum DeckWidgetAction: Equatable, Sendable {
    case openView(viewId: String)
    case openPullRequest(viewId: String, owner: String, repository: String, number: UInt64)
    case refresh(viewId: String)
    case resolveEvent(eventId: String)

    public var url: URL? {
        var components = URLComponents()
        components.scheme = "https"
        components.host = "deli.dev"
        components.path = DevHudWidgetContract.appLinkPath
        switch self {
        case .openView(let viewId):
            components.queryItems = [.init(name: "action", value: "open-view"), .init(name: "view", value: viewId)]
        case .openPullRequest(let viewId, let owner, let repository, let number):
            components.queryItems = [
                .init(name: "action", value: "open-pr"), .init(name: "view", value: viewId),
                .init(name: "owner", value: owner), .init(name: "repository", value: repository),
                .init(name: "number", value: String(number)),
            ]
        case .refresh(let viewId):
            components.queryItems = [.init(name: "action", value: "refresh"), .init(name: "view", value: viewId)]
        case .resolveEvent(let eventId):
            components.queryItems = [.init(name: "action", value: "resolve-event"), .init(name: "event", value: eventId)]
        }
        return components.url
    }
}

public enum DeckNotificationPolicy {
    public static func payloadEventId(_ userInfo: [AnyHashable: Any]) -> String? {
        guard Set(userInfo.keys.compactMap { $0 as? String }) == ["eventId"],
              let value = userInfo["eventId"] as? String,
              value.range(of: "^[A-Za-z0-9_-]{16,128}$", options: .regularExpression) != nil
        else { return nil }
        return value
    }

    public static func text(detailedText: String?, localDetailEnabled: Bool) -> String {
        guard localDetailEnabled, let detailedText, !detailedText.isEmpty else {
            return DevHudWidgetContract.genericNotificationText
        }
        return detailedText
    }
}
